### ISSUE TITLE
Jumplist / Recents launches in taskbar

### ISSUE THREAD

**Issue by edgarogh** [PrismLauncher/PrismLauncher#559]
### Role

I'm an end-user on a Linux Mint desktop.

### Suggestion

I want Prism to have a "jumplist" in the taskbar to launch instances.

### Benefit

This would allow me to launch instances quickly from the Desktop, saving 2 clicks and a new window opening, and would improve OS integration.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

_This suggestion is mostly based on my experience with Linux as I don't have any with Windows or MacOS, but making such a feature multi-platform would probably be welcomed by everybody._

### Implementation ideas

There are two relatively standard ways I can think of to implement such a feature.

The first one, used by Steam in my screenshot, relies on the [Freedesktop Desktop Entry spec](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#extra-actions) which allows `.desktop` files to specify additional "actions" statically. The problem, as just stated, is that these are supposed to be static, and keeping them in sync with the Prism instances would probably not be very clean. The `%k` Exec variable could allow Prism to know from which `.desktop` file it was launched (through a not-so-documented `--desktop-entry-file` argument), and only update this one when there is one, but I don't know if applications are really supposed to dynamically edit their desktop entries anyway.

![image](https://user-images.githubusercontent.com/46636609/205896996-31d972e9-4767-4bab-8314-d67fcea58f2c.png)

---

The second one is probably much cleaner, but requires an additional click (at least on my DE, Cinnamon) and might not work on all desktop environments. It also requires GTK, at least as an optional compilation feature. It works by inserting "recent" entries into the [`GtkRecentManager`](https://docs.gtk.org/gtk3/class.RecentManager.html). The entries have the following structure:

```c
struct GtkRecentData {
  gchar* display_name;
  gchar* description;
  gchar* mime_type;
  gchar* app_name;
  gchar* app_exec;
  gchar** groups;
  gboolean is_private;
}
```
...and are registered alongside a custom URI.

The issues of this method, apart from the new library requirement, is that:
  * This requires Prism to support a URI during launch (maybe as a value to `--launch`, or as an entirely new optional positional first argument). This could however allow more integration with other stuff :shrug: (especially if it supports query parameters to alter the launch options)
  * Not all desktop environments will display the "recents" in this menu, I suppose. It's just a nice Cinnamon feature.
  * It isn't possible to change the icon of entries, AFAIK

![image](https://user-images.githubusercontent.com/46636609/205895683-980c9ecf-846a-4ef4-973b-869c281f0806.png)

**Comment by ghost** [comment]
Came here to make a new issue about that, but for Windows. Being able to select recent instances directly from taskbar jumplist would be good.

**Comment by edgarogh** [comment]
I've spent a good full day experimenting with the techniques I described for Linux/FreeDesktop and it's not really going anywhere. All solutions have major drawbacks, and we desperately need some kind of FreeDesktop standardization if we want to support many desktop environments, which won't happen soon.

Apparently Qt [wants to add a cross platform way to add jumplists](https://bugreports.qt.io/browse/QTBUG-94007) to apps, which would at least solve the issue for Windows and macOS, but this linked issue has been stale for 2+ years. I think we'll have to do the work ourselves.

Regarding Windows @MusaSSH, I don't usually use Windows but for the sake of moving forward on this feature request, I might give it a go. I need to find which Windows API allows us to control custom jumplist entries, specifically from C++, as Windows' APIs tend to be made in C#, from memory. If I can implement that, it'd be a good start before considering other implementations. At least Windows provides this feature explicitly, unlike Linux DEs where it may or may not be implemented on a case by case basis.

**Comment by electricsteve** [comment]
I want this really bad

**Comment by Ryex** [comment]
ok so. for linux we should be writing to 

`$XDG_USER_DATA/recently-used.xbel ` or   `~/.local/share/recently-used.xbel ` if `$XDG_USER_DATA` isn't present in the environment
https://www.freedesktop.org/wiki/Specifications/desktop-bookmark-spec/

optionally KDE uses it's own activity manager  that we could link? probably best not to though
https://api.kde.org/legacy/plasma/plasma-activities/html/classKActivities_1_1Consumer.html

I just did a test and editing the file made KDE pick up the new item

**Comment by edgarogh** [comment]
That's so much better than linking GTK, thank you for finding this! 😮‍💨
I might try to implement this but don't count too much on me. Also I wiped Windows from my PC after it corrupted my Linux partitions (which cost me a new 150€ SSD to fix), so I won't be able to try to implement a cross-platform fix all by myself anymore.

**Comment by edgarogh** [comment]
Here's what the XML entry could look like:

```xml
  <bookmark href="prismlauncher://launch/Liminal Industries" added="2025-12-10T11:41:10.736013Z" modified="2025-12-10T11:41:10.736016Z" visited="2025-12-10T11:41:10.736013Z">
    <title>Liminal Industries</title>
    <desc>Play Liminal Industries with Prism Launcher</desc>
    <info>
      <metadata owner="http://freedesktop.org">
        <mime:mime-type type="x-scheme-handler/prismlauncher"/>
        <bookmark:groups>
          <bookmark:group>Multimedia</bookmark:group>
        </bookmark:groups>
        <bookmark:applications>
          <bookmark:application name="prismlauncher" exec="&apos;prismlauncher %u&apos;" modified="2025-12-10T11:41:10.736014Z" count="1"/>
        </bookmark:applications>
        <bookmark:icon type="image/png" href="file:///home/edgar/.local/share/PrismLauncher/instances/Liminal Industries/minecraft/icon.png" name="org.prismlauncher.PrismLauncher"/>
      </metadata>
    </info>
  </bookmark>
```

A few notes:
* I wrote `prismlauncher://launch/Liminal Industries` as a placeholder but it isn't implemented. It looks like the `prismlauncher://` scheme is only used for MSA OAuth2 (I found `BuildConfig.LAUNCHER_APP_BINARY_NAME + "://oauth/microsoft";` in the code).
* The bookmark group "Multimedia" is not ideal, but [the list of supported groups](https://www.freedesktop.org/wiki/Specifications/desktop-bookmark-spec/#appendixb:registeredgroupnames) doesn't include anything for games specifically.

Other notes that might be related to Cinnamon, but it would be great to check with another DE:
* On Cinnamon, the `<desc>` doesn't do anything
* The icon doesn't work on Cinnamon. Neither the png icon nor the symbolic fallback icon (copied from Prism's `.desktop` file)
* On Cinnamon the recent entries don't show up in Prism's jumplist (when right-clicking the taskbar icon). They can be searched in the Start Menu, but that the only place where I could find them. I think that Cinnamon only shows a recents jumplist for applications that have been marked as "default" for a few hardcoded MIME types (audio, text, video). Other apps don't get this privilege. This is very unfortunate and makes this whole solution ineffective for my use case.

# TLDR:
* The blocker right now is the lack of a way to start an instance from a URI.
* The Freedesktop/XBEL solution actually doesn't do what I want on Cinnamon, which is the DE I use. I probably won't spend time implementing a fix for DEs I don't even use.
